### PR TITLE
SE: Enumerable ContentHash for ordered and unordered sets

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Part of #6964

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/Helpers/HashCodeTest.cs [hunk 1]
```diff
@@ -56,9 +58,77 @@ public void DictionaryContentHash_StableForImmutableDictionary()
...
-            var hashCode2 = SonarAnalyzer.Helpers.HashCode.DictionaryContentHash(dict2);
+            var hashCode1 = HashCode.DictionaryContentHash(dict1);
+            var hashCode2 = HashCode.DictionaryContentHash(dict2);
             hashCode1.Should().Be(hashCode2);
         }
+
+        [TestMethod]
+        public void EnumerableUnorderedContentHash_Empty()
+        {
+#pragma warning disable CA1825 // Avoid zero-length array allocations
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Helpers/HashCodeTest.cs:69]: I'd be fine with disabling this at the beginning of this file once. And not enabling it.

File: analyzers/src/SonarAnalyzer.Common/Helpers/HashCode.cs [hunk 2]
```diff
@@ -28,12 +28,16 @@ public static class HashCode    // Replacement for System.HashCode that is avail
         private const uint PreMultiplier = 3266489917U;
         private const uint PostMultiplier = 668265263U;
         private const int RotateOffset = 17;
+        private const int IntSeed = 393241;
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/HashCode.cs:31]: This is givin "Empty" some random value? Why not keeping `0` for empy? It's valid integer has value like any other

File: analyzers/src/SonarAnalyzer.Common/Helpers/HashCode.cs [hunk 3]
```diff
@@ -28,12 +28,16 @@ public static class HashCode    // Replacement for System.HashCode that is avail
...
         public static int DictionaryContentHash<TKey, TValue>(IDictionary<TKey, TValue> dictionary) =>
             dictionary.Aggregate(0, (seed, kvp) => seed ^ Combine(kvp.Key, kvp.Value));
 
-        public static int EnumerableContentHash<TValue>(IEnumerable<TValue> enumerable) =>
-            enumerable.Aggregate(0, (seed, x) => Combine(seed, x));
+        public static int EnumerableUnorderedContentHash<TValue>(IEnumerable<TValue> enumerable) =>
+            enumerable.Aggregate(IntSeed, (seed, x) => seed ^ (x?.GetHashCode() ?? 0));
+
+        public static int EnumerableOrderedContentHash<TValue>(IEnumerable<TValue> enumerable) =>
+            enumerable.Aggregate(IntSeed, Combine);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/HashCode.cs:40]: For these two, I'd consider single-line documentation comments that will explain the different usage.

File: analyzers/src/SonarAnalyzer.Common/Helpers/HashCode.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/HashCode.cs:31]: I need answers to these questions before approving.

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
There was one missed thread above that needs answers, or to be addressed

File: analyzers/src/SonarAnalyzer.Common/Helpers/HashCode.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/HashCode.cs:31]: Damit! Always the same mistake:
image

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**